File: course1/week3/2_max_value_loot/java/src/Main.java
/**
 *
 * Java implementation to find the most valuable combination of items using a fractional knapsack
 *
 * (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
 *
 **/

import java.text.DecimalFormat;
import java.util.Collections;
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Iterator;

public class Main {

    public static class Item implements Comparable< Item >{
        int V = 0,
            W = 0;
        public Item( int value, int weight ){
            super();
            V = value;
            W = weight;
        }
        public double ratio(){
            return( W > 0 )? ( double ) V  / W : 0;
        }
        public int compareTo( Item rhs ){
            return( ratio() < rhs.ratio() )?  1
                : ( ratio() > rhs.ratio() )? -1
                : ( 0 );
        }
    }

    public static double maxValue( ArrayList< Item > items, int W ){
        double ans = 0;
        Collections.sort( items );
        Iterator< Item > it = items.iterator();
        while( W > 0 && it.hasNext() ){
            Item item = it.next();
            if( item.W <= W ){ // item fits in knapsack
                ans += item.V;
                W -= item.W;
            } else {
                double fraction = ( double ) W / item.W;
                ans += fraction * item.V;
                W = 0;
            }
        }
        return ans;
    }

    public static void main( String[] args ){
        Scanner input = new Scanner( System.in );
        int N = input.nextInt(),
            W = input.nextInt();
        ArrayList< Item > items = new ArrayList<>( Collections.nCopies( N, new Item( 0, 0 ) ) );
        for( int i=0; i < N; ++i ){
            int Vi = input.nextInt(),
                Wi = input.nextInt();
            Item item = new Item( Vi, Wi );
            items.set( i, item );
        }
        double ans = maxValue( items, W );
        System.out.println( new DecimalFormat( "#0.0000" ).format( ans ));
    }
}

File: course1/week3/2_max_value_loot/python/main.py
# python3

##
#
# C++ implementation to find the most valuable combination of items using a fractional knapsack
#
# (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
#
##

from typing import List
from collections import namedtuple

Item = namedtuple( "Item", "V W" )

def comparator( item: Item ) -> float:
    return item.V / item.W

class Solution:
    def maxValue( self, items: List[Item], W: int ) -> float:
        ans = 0
        items.sort( key=comparator, reverse=True )
        for item in items:
            if W == 0:
                break
            if item.W <= W:
                ans += item.V
                W -= item.W
            else:
                fraction = W / item.W
                ans += fraction * item.V
                W = 0
        return ans

if __name__ == '__main__':
    solution = Solution()
    N, W = map( int, input().split() )
    items = []
    for _ in range( N ):
        Vi, Wi = map( int, input().split() )
        item = Item( Vi, Wi )
        items.append( item )
    ans = solution.maxValue( items, W )
    print( "%.4f" % ans )
